runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/DoFnFeatures.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.core.construction;
+
+import org.apache.beam.sdk.state.BagState;
+import org.apache.beam.sdk.state.MapState;
+import org.apache.beam.sdk.state.SetState;
+import org.apache.beam.sdk.state.State;
+import org.apache.beam.sdk.state.ValueState;
+import org.apache.beam.sdk.state.WatermarkHoldState;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignatures;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/**
+ * Features a {@link DoFn} can posses. Each runner might implement a different (sub)set of this
```

Comment:
The `DoFnSignature` is precisely the list of features already, no? It is OK to have helper methods anyhow.

Style point: for a given class, utility static methods usually go in a companion class like `DoFnSignature` (class) & `DoFnSignatures` (utility methods)

Reply:
Agree. The reason I created this class is that I wanted to demonstrate the approach of declaring pipeline requirements and validating that runner supports all requirements. 

That would go as follows:
```java
 Set<Features> required = PipelineFeaturs.extract(pipeline);
 Runner.validateAllFeaturesSupported(required);
```
That way, adding a new feature to pipeline would not require any change to runner core, but the runner would reject the pipeline.

I then realized that this change would be too big and really not related to the annotation, so I returned back to the original approach, where runners explicitly reject features they do not support (and adding unsupported feature needs modification in runners code). I will rename the class as you suggest.

Reply:
Hmmm ... `DoFnSignatures` already exists. It does something different (creates `DoFnSignature` from `DoFn`). Feels weird to mix these two helper functionalities.

Reply:
@kennknowles would you have any suggestions about naming the class? I think this code really should not go to `DoFnSignatures`, can we agree on some alternative name?

Reply:
I quite strongly believe it belongs in `DoFnSignatures` and that what you describe in the rest belongs in `Pipelines`. Static-method-only utility classes tend to be disorganized and undiscoverable. It is better to attach them to the thing that they are most related to.

Reply:
Moved the code there. The reason I didn't like it is that `DoFnSignatures` probably when created served the purpose to create signatures from DoFn, while the code I added makes no use of `DoFnSignature` itself. But there seems to be more code like that already, so I'm fine with that.